sqld/src/query_result_builder.rs

```diff
@@ -0,0 +1,793 @@
+use std::io;
+use std::ops::{Deref, DerefMut};
+
+use anyhow::anyhow;
+use rusqlite::types::ValueRef;
+use serde::Serialize;
+use serde_json::ser::Formatter;
+
+#[derive(Debug, thiserror::Error)]
+#[error(transparent)]
+pub struct QueryResultBuilderError(#[from] anyhow::Error);
+
+impl QueryResultBuilderError {
+    pub fn from_any<E: Into<anyhow::Error>>(e: E) -> Self {
+        Self(e.into())
+    }
+}
+
+/// Identical to rusqlite::Column, with visible fields.
+#[cfg_attr(test, derive(arbitrary::Arbitrary))]
+pub struct Column<'a> {
+    pub(crate) name: &'a str,
+    pub(crate) decl_ty: Option<&'a str>,
+}
+
+impl<'a> From<(&'a str, Option<&'a str>)> for Column<'a> {
+    fn from((name, decl_ty): (&'a str, Option<&'a str>)) -> Self {
+        Self { name, decl_ty }
+    }
+}
+
+impl<'a> From<&'a rusqlite::Column<'a>> for Column<'a> {
+    fn from(value: &'a rusqlite::Column<'a>) -> Self {
+        Self {
+            name: value.name(),
+            decl_ty: value.decl_type(),
+        }
+    }
+}
+
+pub trait QueryResultBuilder: Send + 'static {
+    type Ret: Sized + Send + 'static;
+
+    /// (Re)initialize the builder. This method can be called multiple times.
+    fn init(&mut self) -> Result<(), QueryResultBuilderError>;
+    /// start serializing new step
+    fn begin_step(&mut self) -> Result<(), QueryResultBuilderError>;
+    /// finish serializing current step
+    fn finish_step(
+        &mut self,
+        affected_row_count: u64,
+        last_insert_rowid: Option<i64>,
+    ) -> Result<(), QueryResultBuilderError>;
+    /// emit an error to serialize.
+    fn step_error(&mut self, error: crate::error::Error) -> Result<(), QueryResultBuilderError>;
+    /// add cols description for current step.
+    /// This is called called at most once per step, and is always the first method being called
+    fn cols_description<'a>(
+        &mut self,
+        cols: impl IntoIterator<Item = impl Into<Column<'a>>>,
+    ) -> Result<(), QueryResultBuilderError>;
+    /// start adding rows
+    fn begin_rows(&mut self) -> Result<(), QueryResultBuilderError>;
+    /// begin a new row for the current step
+    fn begin_row(&mut self) -> Result<(), QueryResultBuilderError>;
+    /// add value to current row
+    fn add_row_value(&mut self, v: ValueRef) -> Result<(), QueryResultBuilderError>;
+    /// finish current row
+    fn finish_row(&mut self) -> Result<(), QueryResultBuilderError>;
+    /// end adding rows
+    fn finish_rows(&mut self) -> Result<(), QueryResultBuilderError>;
+    /// finish serialization
+    fn finish(&mut self) -> Result<(), QueryResultBuilderError>;
+    /// returns the inner ret
+    fn into_ret(self) -> Self::Ret;
+    /// Returns a `QueryResultBuilder` that wraps Self and takes at most `n` steps
+    fn take(self, limit: usize) -> Take<Self>
```

Comment:
it's a combinator method and is very common. Actually, you can't overwrite it, because only this module can construct a `Take`. This is a very common pattern, just like `Iterator`, 'Stream' etc

Reply:
I think that the modern pattern is to keep the trait small and implement convenience methods in `Ext` (such as `Stream` vs `StreamExt` or `tokio::AsyncWrite` vs `tokio::AsyncWriteExt`). However, this adds hassle that is probably not justified for this internal trait.